### PR TITLE
Update state transition functions (v0.7.1)

### PR DESCRIPTION
**Pull request by cemozerr**
## PR Description
Updates state transition functions to spec version 0.7.1

### CODE REVIEW

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/Constants.java [hunk 1]
```diff
@@ -30,7 +30,7 @@
   public static int MAX_INDICES_PER_ATTESTATION = 4096; // 2^5
   public static int MIN_PER_EPOCH_CHURN_LIMIT = 4; // 2^2 withdrawals
   public static int CHURN_LIMIT_QUOTIENT = 65536; // 2^2 withdrawals
-  public static int BASE_REWARD_PER_EPOCH = 5; // 2^2 withdrawals
+  public static int BASE_REWARDS_PER_EPOCH = 5; // 2^2 withdrawals
```
Comment (schroedingerscode) [inline ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/Constants.java:33]: Let's remove the 2^2 withdrawals comment. The math doesn't make sense, and I think it's a copy/paste error.

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/BeaconStateUtil.java [hunk 2]
```diff
@@ -23,7 +23,9 @@
 import static tech.pegasys.artemis.datastructures.Constants.LATEST_ACTIVE_INDEX_ROOTS_LENGTH;
 import static tech.pegasys.artemis.datastructures.Constants.LATEST_RANDAO_MIXES_LENGTH;
 import static tech.pegasys.artemis.datastructures.Constants.LATEST_SLASHED_EXIT_LENGTH;
+import static tech.pegasys.artemis.datastructures.Constants.MAX_DEPOSITS;
```
Comment (schroedingerscode) [inline ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/BeaconStateUtil.java:26]: Why was this added if no code was added that relies on it? We should make sure to clean up the static imports.

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/BeaconStateUtil.java [hunk 3]
```diff
@@ -23,7 +23,9 @@
...
 import static tech.pegasys.artemis.datastructures.Constants.MAX_DEPOSIT_AMOUNT;
+import static tech.pegasys.artemis.datastructures.Constants.MAX_EFFECTIVE_BALANCE;
```
Comment (schroedingerscode) [inline ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/BeaconStateUtil.java:28]: Same as above.

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/BeaconStateUtil.java [hunk 4]
```diff
@@ -44,6 +46,8 @@
 import java.util.Objects;
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
+
+import org.apache.logging.log4j.Level;
```
Comment (schroedingerscode) [inline ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/BeaconStateUtil.java:50]: Same as above.

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/blocks/BeaconBlock.java [hunk 5]
```diff
@@ -91,7 +93,7 @@ public boolean equals(Object obj) {
     }
 
     BeaconBlock other = (BeaconBlock) obj;
-    return slot == other.getSlot()
+    return Objects.equals(this.getSlot(), other.getSlot())
```
Comment (dangerousfood) [inline ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/blocks/BeaconBlock.java:96]: Why use this syntax over `this.getSlot().isEquals(other.getSlot())`

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/state/Crosslink.java [hunk 6]
```diff
@@ -47,6 +47,14 @@ public Crosslink(Crosslink crosslink) {
     this.data_root = crosslink.getData_root();
   }
 
+  public Crosslink() {
+    this.shard = UnsignedLong.ZERO;
+    this.start_epoch = UnsignedLong.ZERO;
+    this.end_epoch = UnsignedLong.ZERO;
+    this.parent_root = Bytes32.ZERO;
+    this.data_root = Bytes32.ZERO;
+  }
```
Comment (dangerousfood) [inline ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/state/Crosslink.java:56]: Does this introduce a situation where the application might think that it is on `shard = 0`, `start_epoch = 0`, `end_epoch = 0`, `parent_root = 0` and `data_root = 0`. Basically a situation where the default value will be mistaken for a legitimate value?

**Review by dangerousfood — APPROVED** [review]
lgtm

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/blocks/BeaconBlock.java (hunk 5 above)
Comment (cemozerr) [reply ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/blocks/BeaconBlock.java:96]: I just wanted to make sure it is in line with other equals statements

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/blocks/BeaconBlock.java [hunk 7]
```diff
@@ -25,7 +27,7 @@
 public final class BeaconBlock {
 
   // Header
-  private long slot;
+  private UnsignedLong slot;
```
Comment (jrhea) [inline ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/blocks/BeaconBlock.java:30]: Finally!!!!!

File: ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/BlockProcessorUtil.java [hunk 8]
```diff
@@ -81,632 +100,475 @@
...
-    return state.getSlot().compareTo(UnsignedLong.valueOf(block.getSlot())) == 0;
-  }
-
-  /**
-   * @param state
-   * @param block
-   */
-  public static void process_randao(BeaconState state, BeaconBlock block) {
+  // @v0.7.1
+  public static void process_randao(BeaconState state, BeaconBlockBody body) {
```
Comment (jrhea) [inline ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/BlockProcessorUtil.java]: Why remove the multiline comments?

File: ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/BlockProcessorUtil.java [hunk 9]
```diff
@@ -81,632 +100,475 @@
...
             state
                 .getValidator_registry()
                 .get(toIntExact(proposer_slashing.getProposer_index().longValue()));
 
-        // Verify that the epoch is the same
         checkArgument(
             slot_to_epoch(proposer_slashing.getHeader_1().getSlot())
                 .equals(slot_to_epoch(proposer_slashing.getHeader_2().getSlot())),
-            "Epoch is not the same in process_proposer_slashings");
+            "Verify that the epoch is the same");
```
Comment (jrhea) [inline ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/BlockProcessorUtil.java]: I would prefer if you still keep the name of the method in the checkArgument errorMessage

File: ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/BlockProcessorUtil.java [hunk 10]
```diff
@@ -81,632 +100,475 @@
...
+        boolean slashed_any = false;
+
+        List<UnsignedLong> attesting_indices_1 = attestation_1.getCustody_bit_0_indices();
+        attesting_indices_1.addAll(attestation_1.getCustody_bit_1_indices());
+        List<UnsignedLong> attesting_indices_2 = attestation_2.getCustody_bit_0_indices();
+        attesting_indices_2.addAll(attestation_1.getCustody_bit_1_indices());
+
+        // retainAll is being used to get the intersection of these two lists
+        List<UnsignedLong> sorted_intersection_of_indices = attesting_indices_1;
+        sorted_intersection_of_indices.retainAll(attesting_indices_2);
```
Comment (jrhea) [inline ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/BlockProcessorUtil.java:346]: nice

**Review by jrhea — COMMENTED** [review]
I left a few comments mainly regarding the fact that you removed a ton of nice multiline JavaDoc comments - plz fix this.  I dont want the quality of documentation to go down bc we were hurrying. 

Also, lets always put the name of the calling method in the checkArgument errorMessage.

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/state/Crosslink.java (hunk 6 above)
Comment (cemozerr) [reply ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/state/Crosslink.java:56]: It might, but the spec had a place where it was initializing a new Crosslink object without passing any arguments, so I did this. How can we do this more safely?

File: ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/BlockProcessorUtil.java (hunk 8 above)
Comment (schroedingerscode) [reply ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/BlockProcessorUtil.java]: Same. Even though these aren't the most useful Javadocs, I'd rather have them than not.

File: ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/EpochProcessorUtil.java [hunk 11]
```diff
@@ -74,995 +99,576 @@
...
-      if (a.getData()
-          .getTarget_root()
-          .equals(get_block_root(state, get_epoch_start_slot(get_previous_epoch(state))))) {
-        attestations.add(a);
-      }
-    }
-    return attestations;
+  private static List<PendingAttestation> get_matching_head_attestations(BeaconState state, UnsignedLong epoch)
+          throws IllegalArgumentException {
+    return get_matching_source_attestations(state, epoch).stream().filter(a -> a.getData().getBeacon_block_root().equals(get_block_root_at_slot(state, get_attestation_data_slot(state, a.getData())))).collect(Collectors.toList());
```
Comment (schroedingerscode) [inline ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/EpochProcessorUtil.java:162]: Dude nice use of stream here.

File: ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/EpochProcessorUtil.java [hunk 12]
```diff
@@ -74,995 +99,576 @@
...
-                      .getExit_epoch()
-                      .plus(UnsignedLong.valueOf(Constants.MIN_VALIDATOR_WITHDRAWABILITY_DELAY)))
-          >= 0;
+    // Update effective balances with hysteresis
+    for (int index = 0; index < state.getValidator_registry().size(); index++) {
+      Validator validator = state.getValidator_registry().get(index);
+      UnsignedLong balance = state.getBalances().get(index);
+      long HALF_INCREMENT = Constants.EFFECTIVE_BALANCE_INCREMENT / 2;
+      if (balance.compareTo(validator.getEffective_balance()) < 0
+              || validator.getEffective_balance().plus(UnsignedLong.valueOf(3 * HALF_INCREMENT)).compareTo(balance) < 0) {
```
Comment (schroedingerscode) [inline ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/EpochProcessorUtil.java:631]: I know this is the order of operations as given in the spec, but I have doubts whether the spec is right. Let's think about this one a little more.

**Review by schroedingerscode — APPROVED** [review]
Some small nitpicks and some comments, but you were very thorough overall. Nice job.

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/BeaconStateUtil.java (hunk 2 above)
Comment (cemozerr) [reply ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/BeaconStateUtil.java:26]: Since spotlessApply will just clean them as soon as we're able to get this branch build, I thought I wouldn't put time now to clean them.

File: ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/EpochProcessorUtil.java (hunk 12 above)
Comment (cemozerr) [reply ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/EpochProcessorUtil.java:631]: Agreed. Wanna ask this on gitter?
